Add details on native packaging requirements exposed by mobile platforms (#27)
pypackaging-native/pypackaging-native: docs/key-issues/cross_platform.md

[iOS](https://github.com/freakboy3742/chaquopy/tree/iOS-support/server/pypi))
that is analogous to the tools used by conda-forge to build binary artefacts.

## Problems

There is currently a gap in _communicating target platform details to the
build system_. While a build system like autoconf or CMake may support
cross-platform compilation, and a project may be able to cross-compile binary
artefacts, invocation of a `pyproject.toml` build hook typically assumes that the
platform running the build will be the platform that ultimately runs the Python
code. As a result, `sys.platform`, or the various attributes of the `platform`
and `sysconfig` modules can't be used as part of the build process.

_Running Python code_ for the host (cross) platform is not possible (modulo
using an emulator), but Python packages have not designed for this. to be
avoided. For example, `numpy` and `pybind11` ship headers and have
`get_include()` functions in their main namespaces to obtain the path to those
headers. That is clearly a problem, which packages dependending on those
headers have to work around (often done by patching those packages with
hardcoded paths within a cross-compilation setup).

> `pip` provides limited support for installing binaries for a different platform

Reviewer: @freakboy3742 is this actually a problem? It doesn't seem clear to me that it is. What is the use case here to install packages on a system where you know they won't run?

Reviewer 2: Many packages that use NumPy's C API call `import numpy; numpy.get_include()` to find the location of NumPy's C headers. If you're cross-compiling, and the build platform is different to the NumPy package's platform, then `import numpy` will of course fail. This can be worked around by setting an internal NumPy variable to prevent it from importing its native libraries. But since most Python packages are not written with cross-compilation in mind, they need to be patched to set this variable.

Reviewer 2: ... or to avoid `get_include` altogether and use a hardcoded path instead.

Reviewer 3: So the question is, how does patching the import tie into pip needing to install packages for the cross compile platform.

I think that depending on the package, you don't need this at all. e.g. for pybind11, it is arch-independent (header only) so you can install it any which way and do the `import pybind11; pybind11.get_include()` from the native python executable.

This is a problem for numpy, though, because it also has static libraries. So you do need the right numpy for this. In that case, you do need to be able to pip install the cross-built numpy.

You *also* need to patch it or hardcode the path, since you need its binary artifacts, but don't want to execute its code.

Reviewer: I'm very aware of the issues with `get_include`, which is why I added it in the paragraph above. My question is not about that though, it's about this paragraph (_"`pip` provides limited support ..."_).

Of all the cross-building I've seen, that never once included using `pip` to install the non-native flavor of a build dependency. Is that something you need in BeeWare, and if so could you explain in a bit more detail how this goes wrong?

Reviewer 3: Sure, cross compiling in package manager based ecosystems, such as conda or void-linux or yocto, invariably have their own way to install non-native flavors for packages where the native flavor simply will not cut it.

I would still assume in at least some cases though, that specialized tooling doesn't exist and it's simplest to build the tooling around `pip install`. But it's probably the minority of cases.

Reviewer: `crossenv`'s docs actually do use `pip` to install both build and host packages: https://crossenv.readthedocs.io/en/latest/quickstart.html#make-the-cross-environment. But that seems to work (at least according to the docs). 

The next lines: _"however, these flags only work for the selection of pre-built binary artefacts, and are therefore constrained to the set of platform and ABI tags published by the author."_ hint at what's missing. But I'm missing some concrete scenario/details of where it breaks down. Maybe its more about wheel tags in general that `pip` does not understand?

Author: @rgommers The use case is mobile, web, and embedded systems. The critical difference is that these are systems where you *can't* (or, at the very least, can't without extreme difficulty) run the development environment on the device where the software will be used. For example, if you're building an iPhone app, you building the app on macOS. You can't run gcc (or any other binary tooling) on the device; and Apple will outright reject apps that try to contact an external (non-AppStore) repository to download additional code, so you can't run pip on the device either. That means pip needs to be executed on macOS, targeting iOS.

Reviewer: That much is clear, but the next sentences imply that there is support for this via `--platform` & co. I just tried, and while the UX is pretty bad it took me only a couple of minutes to convince `pip` to install a Windows wheel on my macOS system with:
```
pip install scikit-image --platform=win_amd64 --no-deps --target . --only-binary :all: --implementation=cp
```
So if you replace `win_amd64` with whatever the right tag is for iOS, doesn't that work?

Author: It does - however, it only works if you enable `--only-binary`. In practice, we've found that we can't enable `--only-binary` because there are a non-trivial number of non-binary packages that are only available as source packages, not as noarch wheels. So, in order to invoke pip in a single pass (thereby getting a consistent dependency resolution pass), you have to allow source packages, which then opens the door to pip trying to compile numpy (et al) from source.

Admittedly, this is an edge case, and one that would go away if publishing noarch wheels was something that was required (and/or there was cross-compilation support in pip) - but it's a limitation that exists. Hence: "limited support".

Author: Well... there are 2 possible fixes:

1. Make everyone do the right thing, and fix all historical packages published in the world where wheels weren't required; or
2. Modify pip to deal with this edge case of packaging.

I can't see any world where (1) is the easier solution :-)

Reviewer: (2) isn't an option, because there is unfortunately no way for `pip` to know if the package authors simply neglected to upload a `-none-any` wheel, or if there's compiled code in there.

There are also pure Python packages that don't have wheels because their whole purpose is to run Python code or provide system-dependent metadata at build time, e.g. `oldest-supported-numpy` and `cupy`.

Reviewer: Thinking some more about how this actually plays out in practice. When you're dealing with cross-compiling a large amount of packages, your main concern is that you have to cross-compile every non-pure wheel. For something like iOS there is nothing on PyPI. So you're setting up your own wheelhouse anyway. At best, you can use PyPI as a fallback for `-none-any` wheels. So when those are missing, you treat the package like you do all packages which aren't pure Python, and put a `-none-any` wheel in your own wheelhouse. So this issue is then "for each missing `-none-any` wheel I have to produce that wheel and host it myself". Which is extra work, but probably a minor amount compared to the task of building wheels for non-pure packages.

A significant improvement here would be if `pip install --platform ...` gained the capability to build wheels from sdists. But that already falls under "standardize cross-compiling". The `--platform ...` specifiers would need to be passed down to the build backend, which is only possible after standization.

So I'm thinking of moving the main description of what `pip` can do to the "current status" section, and under "problems" leave only the "every missing `-none-any` wheel incurs some extra effort". WDYT?

Reviewer 2: > A significant improvement here would be if `pip install --platform ...` gained the capability to build wheels from sdists.

Many of these sdists are actually pure-Python, or have a pure-Python fallback if their extension modules fail to compile. So for Android, BeeWare / Chaquopy patches its copy of pip to remove the block on installing sdists when `--platform` is supplied. Instead, we monkey-patch setuptools to immediately throw an exception when it attempts to execute a native compiler. This allows most pure-Python fallbacks to work correctly, while providing a more comprehensible error message if there is no fallback.

@freakboy3742 can confirm this, but I think on iOS we work a bit differently. We monkey-patch the standard library so that pip, running on macOS, believes it's actually running on iOS. That means the `--platform` option is not required, so building sdists is no problem as long as they don't attempt to execute a native compiler.

Either way, we don't need to host none-any wheels anywhere. If the package is able to produce a none-any wheel, then pip will build it automatically.

Reviewer: Thanks. That sounds pragmatic, and similar to what `crossenv` does. Ideally such patching won't be needed anymore in a future that has more solid cross-compilation support. But that does rely on something being standardized.

Author: Confirming that yes, Briefcase on iOS is effectively doing "crossenv-lite"; and yes, the best-case endpoint is that externally hacked-on crossenv-style approaches aren't needed.

# Cross-platform installation

The historical assumption of compilation is that the platform where the code is
compiled will be the same as the platform where the final code will be executed
(if not literally the same machine, then at least one that is CPU and ABI
compatible at the operating system level). This is a reasonable assumption for
most desktop projects; However, for mobile platforms, this isn't the case.

On mobile platforms, an app is compiled on a desktop platform, and transferred
to the mobile device (or a simulator) for testing. The compiler is not executed
on device. Therefore, it must be possible to build a binary artefact for a CPU
architecture and a ABI that is different the platform that is running the
compiler.

> 

Reviewer: Embedded devices like the Raspberry Pi (`armv7` in particular) often need cross-compiling too. That's a prominent enough case to mention here I think (also xref https://www.piwheels.org/ for that).

And in general, cross-compiling to architectures that are not/less available in free CI (`ppc64le`, `s390x`) would be nice to have as well

Author: Good point - I'll add a note about this.

A microcosm of this problem exists on macOS as a result of the Apple Silicon
transition. Most CI systems don't provide native ARM hardware, but most
developers will still want ARM64-compatible build artefacts. Apple has provided
the tools compile [fat binaries](multiple_architectures.md) on x86_64 hardware;
however, in this case, the host platform (macOS on x86_64) will still be one of
the outputs of the compilation process. For mobile platforms, the computer that

> compiles the code will not be able to execute the code that has been compiled.

Reviewer 4: I guess we're all viewing the world through our own little prism, but leading with mobile and using "microcosm" for osx-arm is kinda counter to my experience in Python packaging (even though that might be a very justified POV in beeware).

I think if we add a cross-compilation page (and certain bits and pieces are already sprinkled here and there, e.g. in `background/compilation_concepts.md`), we should aim to have a wider view on the topic.

I agree though that it probably merits its own page. It's a very tricky problem at scale (because so many packages have build instructions that just assume everything runs on native hardware), as we see constantly in conda-forge for osx-arm, linux-aarch, linux-ppc64le, etc.

Author: My framing of this as a "microcosm" is on the basis that the problem is broadly similar, but less complicated. macOS only needs to deal with 2 platforms; one of those platforms will always be the host platform, so the build tooling will run on host; and the compiler can always work in a single pass. For iOS, I've had to deal with up to 5 platforms simultaneously for iOS, and *none* of them were native, so none of the build tools run natively, and it almost always requires multiple compiler passes.

## Potential solutions or mitigations

Compiler and build toolchains (e.g., autoconf/automake) have long supported
cross-compilation; however, these cross-compilation capabilities are easy to
break unless they are exercised regularly.

In the Python space, tools like [crossenv](https://github.com/benfogle/crossenv)
also exist; these tools use a collection of path hacks and overrides of known
sources of platform-specific details (like `distutils`) to provide a
cross-compilation environment. However, these solutions tend to be somewhat

> fragile as aren't first-class citizens of the Python ecosystem.

Reviewer: This is a good point to make. They should become first-class citizens of at least the build backends that support them. I certainly want that for `meson-python` (and it's part of the way there). The rest of the Python tooling doesn't need to know I believe, it's a build system/backend thing.

Author: Yeah - the connection with core Python is entirely a result of `distutils`, which is thankfully no longer a problem in Python 3.12.

to the mobile device (or a simulator) for testing. The compiler is not executed
on device. Therefore, it must be possible to build a binary artefact for a CPU
architecture and a ABI that is different the platform that is running the
compiler.

Cross compilation issues also emerge when dealing with continuous
integration/deployment (CI/CD). CI/CD platforms (such as Github Actions)
generally provide the "common" architectures - often only x86-64 - however, a
project may want to produce binaries for other platforms (e.g., ARM support for
Raspberry Pi devices; PowerPC or s390 for mainframe/server devices; or for
mobile platforms). These binaries won't run natively on the host CI/CD system
(without some sort of emulation); but code can be compiled for the target
platform.

macOS also experiences this as a result of the Apple Silicon transition. Apple
has provided the tools to compile [fat binaries](multiple_architectures.md) on
x86_64 hardware; however, in this case, the host platform (macOS on x86_64) will
still be one of the outputs of the compilation process, and the resulting binary
will run on the CI/CD system.

## Current state

> Native compiler and build toolchains (e.g., autoconf/automake, CMake) have long
> supported cross-compilation; however, these cross-compilation capabilities are
> easy to break unless they are exercised regularly.

Reviewer 4: ```suggestion
Native compiler and build toolchains (e.g., autoconf/automake, CMake) have long
supported cross-compilation; however, such cross-compilation capabilities for any
given project tend to bitrot and break easily unless they are exercised regularly.
```

CPython's build system includes some support for cross-compilation. This support
is largely based on leveraging autoconf's support for cross compilation. This
support wasn't well integrated into distutils and the compilation of the binary
portions of stdlib; however, with the deprecation and removal of disutils in

> Python 3.12, this situation has improved.

Reviewer 4: I'm not sure that we're at "improvement" yet. Removing distutils certainly removes an impediment, but it doesn't mean we suddenly have good cross-compilation support - doubly so for windows, where autoconf is not usually at home, and certainly not for CPython builds.

Author: Fair :-) It's a step on a good path, but there's a way to go. I'll clarify.

The specification of PEP517 means cross-platform compilation support has been
largely converted into a concern for individual build systems to manage.

## Problems

> There is currently a small gap in communicating target platform details to the
> build system. While a build system like autoconf or Cmake may support

Reviewer 4: That's one deceptively small "small" IMO.
```suggestion
There is currently a gap in communicating target platform details to the
build system. While a build system like autoconf or CMake may support
```

cross-platform compilation, and a project may be able to cross-compile binary
artefacts, invocation of the PEP517 build interface currently assumes that the
platform running the build will be the platform that ultimately runs the Python
code. As a result, `sys.platform`, or the various attributes of the `platform`
library can't be used as part of the build process.

> `pip` provides limited support for installing binaries for a different platform
> by specifying a `--platform`, `--implementation` and `--abi` flags; however,
> these flags only work for the selection of pre-built binary artefacts.

Reviewer 4: ```suggestion
`pip` provides limited support for installing binaries for a different platform
by specifying a `--platform`, `--implementation` and `--abi` flags; however,
these flags only work for the selection of pre-built binary artefacts, and are
hence constraint to the set of available platform & ABI tags.
```

## History

Tools like [crossenv](https://github.com/benfogle/crossenv) can be used to trick
Python into performing cross-platform builds. These tools use path hacks and
overrides of known sources of platform-specific details (like `distutils`) to
provide a cross-compilation environment. However, these solutions tend to be
somewhat fragile as they aren't first-class citizens of the Python ecosystem.

[The BeeWare Project](https://beeware.org) also uses a version of these
techniques. On both platforms, BeeWare provides a custom package index that
contains pre-compiled binaries ([Android](https://chaquo.com/pypi-7.0/);
[iOS](https://anaconda.org/beeware/repo)). These binaries are produced using a

> forge-like set of tooling

Reviewer 4: Not sure what "forge-like"means, could you clarify?

([Android](https://github.com/chaquo/chaquopy/tree/master/server/pypi);
[iOS](https://github.com/freakboy3742/chaquopy/tree/iOS-support/server/pypi)).

## Relevant resources

TODO

## Potential solutions or mitigations

> At it's core, what is required is a recognition that cross-platform builds as a
> use case that the Python ecosystem supports.

Reviewer 4: Grammar is off here: "what's required is [...] that [xyz] as a use case". Perhaps:
```suggestion
At it's core, what is required is a recognition that cross-platform builds are a
use case that the Python ecosystem supports.
```

Reviewer 5: also `its`… or maybe `at the core`

In concrete terms, for native modules, this would require either:

1. Extension of the PEP517 interface to allow communicating the desired target
   platform as part of a binary build; or

2. Formalization of the "platform identification" interface that can used by
   PEP517 build backends to identify the target platform, so that tools like
   `crossenv` can provide a reliable proxied environment for cross-platform

>    builds.

Reviewer 4: I think in terms of requirements I'd add (resp. subsume):
3. metadata separation into build & target requirements
4. infrastructure support (e.g. hooks, etc.) to tell the PEP517 interface that certain steps (e.g. code generation) cannot be run on the target hardware

On mobile platforms, an app is compiled on a desktop platform, and transferred
to the mobile device (or a simulator) for testing. The compiler is not executed
on device. Therefore, it must be possible to build a binary artefact for a CPU
architecture and an ABI that is different from the platform that is running the
compiler. The situation is similar for embedded devices.

Cross compilation issues also emerge when dealing with continuous
integration/deployment (CI/CD). CI/CD platforms (such as Github Actions)
generally provide the "common" architectures - often only x86-64 - however, a
project may want to produce binaries for other platforms (e.g., ARM support for
Raspberry Pi devices; PowerPC or s390x for mainframe/server devices; or for
mobile platforms). These binaries won't run natively on the host CI/CD system
(without some sort of emulation, for example with QEMU); but code can be
compiled for the target platform.

macOS also experiences this as a result of the Apple Silicon transition. Apple
has provided the tools to make cross compilation from x86-64 to arm64 as easy
as possible, as well as to compile [fat binaries](multiple_architectures.md)

> (supporting x86-64 and arm64 at the same time) on x86-64 hardware. In the latter

Author: FWIW: You can compile fat binaries on M1 hardware as well - it's isn't x86_64-specific.

Reviewer: Good catch, I'll fix that up.
